File: 05_Bagels/kotlin/bagels.kt
import java.util.Random

fun main() {
    println("BAGELS")
    println("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    var a = mutableListOf(0, 0, 0)
    var y = 0
    var t = 255
    println()
    println()
    println()
    println("WOULD YOU LIKE THE RULES (YES OR NO)? ")
    var astring = readLine()
    if (astring?.substring(0, 1) != "N") {
        println()
        println()
        println("I AM THINKING OF A THREE-DIGIT NUMBER.  TRY TO GUESS")
        println("MY NUMBER AND I WILL GIVE YOU CLUES AS FOLLOWS:")
        println("   PICO   - ONE DIGIT CORRECT BUT IN THE WRONG POSITION")
        println("   FERMI  - ONE DIGIT CORRECT AND IN THE RIGHT POSITION")
        println("   BAGELS - NO DIGITS CORRECT")
    }
    var i = 0
    var random = Random()
    while (i < 3) {
        a[i] = random.nextInt(9) + 1
        if (i > 0) {
            for (j in 0 until i) {
                if (a[i] == a[j])
                {
                    a[i] = random.nextInt(9) + 1
                    i--
                }
            }
        }
        i++
    }
    println()
    println("O.K.  I HAVE A NUMBER IN MIND.")
    do {
        var i = 1
        while (i <= 20) {
            do {
                var repeatGuess = false
                println("GUESS #" + i)
                astring = readLine()
                if (astring?.length != 3) {
                    println("TRY GUESSING A THREE-DIGIT NUMBER.")
                    repeatGuess = true
                } else {
                    for (j in 0..2) {
                        if (astring[j].toInt() < 48 || astring[j].toInt() > 57) {
                            println("WHAT?")
                            repeatGuess = true
                        }
                    }

                    if (astring[0] == astring[1] ||
                        astring[1] == astring[2] ||
                        astring[2] == astring[0]
                    ) {
                        println("OH, I FORGOT TO TELL YOU THAT THE NUMBER I HAVE IN MIND")
                        println("HAS NO TWO DIGITS THE SAME.")
                        repeatGuess = true
                    }
                }
            } while (repeatGuess);

            var c = 0
            var d = 0
            for (j in 0..2) {
                if (astring!![j]?.toInt() - 48 == a[j]) {
                    c++
                    d++
                }
                if (j < 2 && astring!![j]?.toInt() - 48 == a[j + 1]) c++
            }
            if (astring!![2]?.toInt() - 48 == a[0]) c++
            if (astring!![0]?.toInt() - 48 == a[2]) c++

            if (d == 3) {
                break;
            }
            if (c != 0) {
                for (j in 1..c) {
                    print("PICO ")
                }
            }
            if (d != 0) {
                for (j in 1..d) {
                    print("FERMI ")
                }
            }
            if (c + d == 0) {
                print("BAGELS ")
            }
            println()
            i++
        }

        if (i == 21) {
            println("OH WELL.")
            println("THAT'S TWENTY GUESSES.  MY NUMBER WAS " + a[0] + a[1] + a[2])
        }
        else {
            println("YOU GOT IT!!!")
            println()
            y++
        }
        println("PLAY AGAIN (YES OR NO)?")
        astring = readLine()
    } while (astring?.substring(0, 1) == "Y");
    if (y != 0) {
        println()
        println("A $y POINT BAGELS BUFF!!")
    }
    println ("HOPE YOU HAD FUN.  BYE.")
}

File: 05_Bagels/lua/Bagels.lua
---
--- Bagels
--- Ported by Joe Nellis.
--- Text displayed is altered slightly from the original program to allow for
--- more (or less) than three digits to be guessed. Change the difficulty to
--- the number of digits you wish in the secret code.
---

--- difficult is number of digits to use
local difficulty = 3

print [[
                                BAGELS
              CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY

]]

function getInput(prompt)
    io.write(prompt)
    io.flush()
    local input = io.read("l") 
    if not input then  --- test for EOF
        print("GOODBYE")
        os.exit(0)
    end
    return input
end

local needsRules = getInput("WOULD YOU LIKE THE RULES? (YES OR NO) ")
print()
if needsRules:match("[yY].*") then
    print(string.format( [[
I AM THINKING OF A %u DIGIT NUMBER.  TRY TO GUESS
MY NUMBER AND I WILL GIVE YOU CLUES AS FOLLOWS:
   PICO   - ONE DIGIT CORRECT BUT IN THE WRONG POSITION
   FERMI  - ONE DIGIT CORRECT AND IN THE RIGHT POSITION
   BAGELS - NO DIGITS
   ]], difficulty))
end

function play(numDigits, score)
    local digits = { "1", "2", "3", "4", "5", "6", "7", "8", "9", "0" }
    --- secret number must not have duplicate digits
    --- randomly swap numDigits at the head of this list to create secret number
    for i = 1, numDigits do
        local j = math.random(1, 10)
        digits[i], digits[j] = digits[j], digits[i]
    end

    print "O.K.  I HAVE A NUMBER IN MIND."
    for guessNum = 1, 20 do
        :: GUESS_AGAIN :: ---<!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

        local guess = getInput(string.format("GUESS #%u\t?",guessNum))
        if #guess ~= numDigits then
            print("TRY GUESSING A", numDigits, "DIGIT NUMBER.")
            goto GUESS_AGAIN
        elseif not guess:match("^(%d+)$") then
            print("WHAT?")
            goto GUESS_AGAIN
        else
            --- check if user has duplicate digits
            for i = 1, numDigits - 1 do
                for j = i + 1, numDigits do
                    if (guess:sub(i, i) == guess:sub(j, j)) then
                        print("OH, I FORGOT TO TELL YOU THAT THE NUMBER I HAVE")
                        print("IN MIND HAS NO TWO DIGITS THE SAME.")
                        goto GUESS_AGAIN
                    end
                end
            end
        end

        local report = ""
        --- check for picos, right digit, wrong place
        for i = 1, numDigits do
            for j = i + 1, numDigits - 1 + i do
                if (guess:sub(i, i) == digits[(j - 1) % numDigits + 1]) then
                    report = report .. "PICO "
                end
            end
        end
        --- check for fermis, right digit, right place
        for i = 1, numDigits do
            if (guess:sub(i, i) == digits[i]) then
                report = report .. "FERMI "
            end
        end

        if (report == string.rep("FERMI ", numDigits)) then
            print "YOU GOT IT!!!"
            print ""
            score = score + 1
            goto PLAY_AGAIN
        end
        if (report == "") then
            print("BAGELS")
        else
            print(report)
        end
    end
    print "OH WELL."
    print("THAT'S TWENTY GUESSES. MY NUMBER WAS "
            .. table.concat(digits, "", 1, numDigits))

    :: PLAY_AGAIN :: ---<!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    local playAgain = getInput("PLAY AGAIN (YES OR NO)?")
    print()
    if (playAgain:match("[yY].*")) then
        return play(numDigits, score)
    else
        if (score > 0) then
            print("A", score, "POINT BAGELS BUFF!!")
        end
        print "HOPE YOU HAD FUN. BYE."
    end
end

play(difficulty, 0) --- default is numDigits=3, score=0

